kernel/src/log_segment.rs

```diff
@@ -213,17 +218,144 @@ impl LogSegment {
             .read_json_files(&commit_files, commit_read_schema, meta_predicate.clone())?
             .map_ok(|batch| (batch, true));
 
-        let checkpoint_parts: Vec<_> = self
-            .checkpoint_parts
+        let checkpoint_stream = Self::create_checkpoint_stream(
+            engine,
+            checkpoint_read_schema,
+            meta_predicate,
+            self.checkpoint_parts.clone(),
+            self.log_root.clone(),
+        )?;
+
+        Ok(commit_stream.chain(checkpoint_stream))
+    }
+
+    /// Returns an iterator over checkpoint data, processing sidecar files when necessary.
+    ///
+    /// By default, `create_checkpoint_stream` checks for the presence of sidecar files, and
+    /// reads their contents if present. Checking for sidecar files is skipped if:
+    /// - The checkpoint is a multi-part checkpoint
+    /// - The checkpoint read schema does not contain the add action
+    ///
+    /// For single-part checkpoints, any referenced sidecar files are processed. These
+    /// sidecar files contain the actual add actions that would otherwise be
+    /// stored directly in the checkpoint. The sidecar file batches are chained to the
+    /// checkpoint batch in the top level iterator to be returned.
+    fn create_checkpoint_stream(
+        engine: &dyn Engine,
+        checkpoint_read_schema: SchemaRef,
+        meta_predicate: Option<ExpressionRef>,
+        checkpoint_parts: Vec<ParsedLogPath>,
+        log_root: Url,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + Send> {
+        let need_add_actions = checkpoint_read_schema.contains(ADD_NAME);
+        require!(
+            !need_add_actions || checkpoint_read_schema.contains(SIDECAR_NAME),
+        Error::invalid_checkpoint(
+                "If the checkpoint read schema contains file actions, it must contain the sidecar column"
+            )
+        );
```

Comment:
I'm a little torn - I understand the need for having different read schemas but it seems like a rather leaky abstraction to have sidecars be required for the caller if they need adds. Curious if we think we could encapsulate this here? Maybe have a way to create checkpoint streams for P+M and for adds if we have relatively constrained call sites?

That being said, if we want to go this route don't we need to check for sidecars if the user wants removes as well?